### DISCUSSION TITLE
delete modal function deletes from the first item to the last

### DISCUSSION THREAD

**Discussion by kisa5** [neuefische/questions#490] [Web] [answered]
Hello, since I work with a modal pop up for my delete function, it is not working properly. When I click on the delete button it starts deleting the projects that I added first in order. Before implementing the modal, the delete function were working right. So the delete function is not the problem.

Here's my code from my modal component:

```
export default function Modal({ appear, children }) {
  const modalContent = appear ? <StyledModal>{children}</StyledModal> : null;

  return ReactDOM.createPortal(
    modalContent,
    document.getElementById("modal-root")
  );
}
```
Here the code for the modal delete:

```
export default function ModalDelete({
  appearModalDeleteProject,
  handleDeleteProject,
  handleClose,
}) {
  function deleteAndClose() {
    handleDeleteProject();
    handleClose();
  }
  return (
    <Modal appear={appearModalDeleteProject}>
      <StyledCard>
        <Question>Do you really want to delete your project?</Question>
        <Wrapper>
          <StyledButton onClick={handleClose}>No</StyledButton>
          <StyledButton onClick={deleteAndClose}>Yes</StyledButton>
        </Wrapper>
      </StyledCard>
    </Modal>
  );
} 
```
And here the code where I hand it over: 
```
import dynamic from "next/dynamic";

const ModalDeleteProject = dynamic(() =>
  import("@/components/ModalDeleteProject")
);

export default function ProjectList({ projects, handleDeleteProject }) {
  const [searchTerm, setSearchTerm] = useState("");
  const [appearModalDeleteProject, setAppearModalDeleteProject] = useState();

  return (
    <Wrapper>
      <StyledList>
        <SearchBar
          type="text"
          placeholder="search"
          onChange={(event) => {
            setSearchTerm(event.target.value);
          }}
        />
        {projects
          .filter((val) => {
            if (searchTerm == "") {
              return val;
            } else if (
              val.title.toLowerCase().includes(searchTerm.toLowerCase())
            ) {
              return val;
            }
          })
          .map((project, index) => (
            <ProjectItem key={index}>
              <Link href={`/project/${project.id}`}>{project.title}</Link>
              <DeleteButton
                type="button"
                onClick={() =>
                  setAppearModalDeleteProject(!appearModalDeleteProject)
                }
              >
                x
                <ModalDeleteProject
                  appearModalDeleteProject={appearModalDeleteProject}
                  handleDeleteProject={() => handleDeleteProject(project.id)}
                  handleClose={() => {
                    setAppearModalDeleteProject(false);
                  }}
                />
              </DeleteButton>
            </ProjectItem>
          ))}
      </StyledList>
    </Wrapper>
  );
}
```

Would be grateful for help 👻

**Comment by doemser** [comment]
Most important part for me would be the `handleDeleteProject` function, which I am missing in the code you provided.

**Reply by doemser** [reply]
```javascript

function deleteAndClose() {
    handleDeleteProject();
    handleClose();
  }

```

I guess you need to pass the `project.id` to handleDeleteProject().

If in doubt, please send a current PR or Branch

**Reply by doemser** [reply]
I seem to be wrong, it is very hard to follow your code, prefer to send a PR in the future. An important step in the right direction would be to display code formatted:

```javascript

console.log("this has syntax highlighting");


```

**Reply by doemser** [reply]
```javascript

 <ModalDeleteProject
                  appearModalDeleteProject={appearModalDeleteProject}
                  handleDeleteProject={() => handleDeleteProject(project.id)}
                  handleClose={() => {
                    setAppearModalDeleteProject(false);
                  }}
                />

```

is nested in the delete button. Could the problem find its root there?

**Comment by kisa5** [comment]
Thanks for the reply first of all. Here my [PR](https://github.com/kisa5/capstone-project/pull/15) 🙂 Tried to pass over the project.id like suggested, but this did not work.

**Reply by doemser** [reply]
You are already passing the product.id here:

```javascript
<ModalDeleteProject
                  appearModalDeleteProject={appearModalDeleteProject}
                  handleDeleteProject={() => handleDeleteProject(project.id)}
                  handleClose={() => {
                    setAppearModalDeleteProject(false);
                  }}
                />

```

 I will check your PR

**Comment by doemser** [comment, marked as answer]
This state:

```javascript
 const [appearModalDeleteProject, setAppearModalDeleteProject] = useState();

```

is used for all of your modals. So if you set it true all modals will be displayed stacked on top of each other, and therefore the last one is on top.

Try to move the decision whether a modal is open or not into the modal component itself, so that this state is only valid for the respective modal. This will probably require a slightly more elaborate rebuild of your logic.

**Reply by doemser** [reply]
If you take it right it is quite disadvantageous to map a modal for every project. It would be better to keep one modal and fill it accordingly just in the case a user wants to delete a project.

**Reply by doemser** [reply]
Please keep this answer as the correct answer because it addresses the symptom described. Treating the cause is not part of the question and answering it is not the answer of the thread.

**Comment by shebtastic** [comment]
You could introduce a different state, that manages the modal globally and also serves as a memory of what project you are planning to delete. If you close the modal and decide not to delete any project, we also close the modal and reset the `project.id` saved.

```js
import dynamic from "next/dynamic";

const ModalDeleteProject = dynamic(() =>
  import("@/components/ModalDeleteProject")
);

export default function ProjectList({ projects, handleDeleteProject }) {
  const [searchTerm, setSearchTerm] = useState("");
  const [projectToDelete, setProjectToDelete] = useState(null);

  return (
    <Wrapper>
      <StyledList>
        <SearchBar
          type="text"
          placeholder="search"
          onChange={(event) => {
            setSearchTerm(event.target.value);
          }}
        />
        {projects
          .filter((val) => {
            if (searchTerm == "") {
              return val;
            } else if (
              val.title.toLowerCase().includes(searchTerm.toLowerCase())
            ) {
              return val;
            }
          })
          .map((project, index) => (
            <ProjectItem key={index}>
              <Link href={`/project/${project.id}`}>{project.title}</Link>
              <DeleteButton
                type="button"
                onClick={() =>
                  setProjectToDelete(project.id)
                }
              >
                x
                
              </DeleteButton>
            </ProjectItem>
          ))}
      </StyledList>
      <ModalDeleteProject
                  appearModalDeleteProject={projectToDelete}
                  handleDeleteProject={() => handleDeleteProject(projectToDelete)}
                  handleClose={() => {
                    setProjectToDelete(null)
                  }}
    />
    </Wrapper>
  );
}
```

**Reply by kisa5** [reply]
This is the error message that I get:
<img width="1012" alt="Bildschirmfoto 2023-02-03 um 12 09 59" src="https://user-images.githubusercontent.com/116880659/216589041-1a150440-e02d-4355-91ff-3fdce1e68037.png">

**Reply by shebtastic** [reply]
The rest of the errors would have helped, but I think the issue lies in the modal itself.

https://github.com/kisa5/capstone-project/pull/15/files#diff-3948a2078ae26d26cb494db5eedaec7d22527e3bbf91d2c801a9619bc6bb9847R9

Your modals portal uses `document.getElementById("modal-root")`. On the server side redering phase `document` does not exist, since this is part of the browser API.
Instead of creating a `div` outside your apps DOM-tree, you could render the `div` anywhere and use positioning and `z-index` to make it an overlay. 

If you want to stick with the current implementation, which would be easier, you have to make sure that all imports of the Modal component are wrapped in `next/dynamic`.
